build.rs

```diff
@@ -76,6 +76,89 @@ impl FromStr for PythonInterpreterKind {
     }
 }
 
+trait GetPrimitive {
+    fn get_bool(&self, key: &str) -> Result<bool>;
+    fn get_numeric<T: FromStr>(&self, key: &str) -> Result<T>;
+}
+
+impl GetPrimitive for HashMap<String, String> {
+    fn get_bool(&self, key: &str) -> Result<bool> {
+        match self
+            .get(key)
+            .map(|x| x.as_str())
+            .ok_or(format!("{} is not defined", key))?
+        {
+            "1" | "true" | "True" => Ok(true),
+            "0" | "false" | "False" => Ok(false),
+            _ => Err(format!("{} must be a bool (1/true/True or 0/false/False", key).into()),
+        }
+    }
+
+    fn get_numeric<T: FromStr>(&self, key: &str) -> Result<T> {
+        self.get(key)
+            .ok_or(format!("{} is not defined", key))?
+            .parse::<T>()
+            .map_err(|_| format!("Could not parse value of {}", key).into())
+    }
+}
+
+struct CrossCompileConfig {
+    lib_dir: PathBuf,
+    include_dir: Option<PathBuf>,
+    version: Option<String>,
+    os: String,
+    arch: String,
+}
+
+impl CrossCompileConfig {
+    fn both() -> Result<Self> {
+        Ok(CrossCompileConfig {
+            include_dir: Some(CrossCompileConfig::validate_variable(
+                "PYO3_CROSS_INCLUDE_DIR",
+            )?),
+            ..CrossCompileConfig::lib_only()?
+        })
+    }
+
+    fn lib_only() -> Result<Self> {
+        Ok(CrossCompileConfig {
+            lib_dir: CrossCompileConfig::validate_variable("PYO3_CROSS_LIB_DIR")?,
+            include_dir: None,
+            os: env::var("CARGO_CFG_TARGET_OS").unwrap(),
+            arch: env::var("CARGO_CFG_TARGET_ARCH").unwrap(),
+            version: env::var_os("PYO3_PYTHON_VERSION").map(|s| s.into_string().unwrap()),
```

Comment:
As it's only used when cross compiling, let's keep that explicit:

```suggestion
            version: env::var_os("PYO3_CROSS_PYTHON_VERSION").map(|s| s.into_string().unwrap()),
```